File: 21553 - Sai Nandan D/Week 5/912_Sort an Array.py
class Solution:
    def sortArray(self, nums: List[int]) -> List[int]:        
        
        def max_heapify(n, i):
            l = 2*i+1
            r = l+1
            
            max_i = i
            
            if l < n and nums[l] > nums[max_i]:
                max_i = l            
                
            if r < n and nums[r] > nums[max_i]:
                max_i = r
                
            if max_i != i:
                nums[max_i], nums[i] = nums[i], nums[max_i]
                max_heapify(n, max_i)
                
        n = len(nums)
        
        for i in range(n//2+1)[::-1]: 
            max_heapify(n, i) 

        for i in range(n)[::-1]: 
            nums[i], nums[0] = nums[0], nums[i]
            max_heapify(i, 0)
            
        return nums

File: 21561 - Suvam Tamang/Week 5/912. Sort an Array.java
class Solution {
    public int[] sortArray(int[] nums) {
        int N = nums.length;
        
        for(int i=(N/2)-1; i>=0; i--)
            heapify(nums, nums.length, i);
        
        for(int i = N-1; i>=1; i--) {
            int temp = nums[i];
            nums[i] = nums[0];
            nums[0] = temp;
            
            heapify(nums, i, 0);
        }
        return nums;
    }
    
    public void heapify(int[] nums, int N, int i) {
        int largest = i;
        int left = 2*i + 1;
        int right = 2*i + 2;
        
        if(left < N && nums[left] > nums[largest])
            largest = left;
        
        if(right < N && nums[right] > nums[largest])
            largest = right;
        
        if(largest != i) {
            int temp = nums[largest];
            nums[largest] = nums[i];
            nums[i] = temp;
            
            heapify(nums, N, largest);
        }
    }
}
